doc/md: add domain testing doc (#2963)
ariga/atlas: doc/md/guides/testing/testing-domains.mdx

1. A valid US Postal Code.
2. And invalid US Postal Code.

```hcl title="schema.test.hcl"
test "schema" "postal" {
  parallel = true
  exec {
    sql = "select '12345'::us_postal_code"
  }
  catch {
    sql = "select 'hello'::us_postal_code"
  }
}
```

This test uses the [`catch`](/testing/schema#catch-command) command which expects
the SQL statement to fail. In this case, we expect "hello" to fail as a valid ZIP code.

:::info
To learn more about how to write tests, view the [testing docs](/testing/schema).
:::

To run this test, run:

```
atlas schema test --env dev
```

The output should be similar to:

```testoutput title="Test Output"
-- PASS: domains (3ms)
PASS
```

### Table Driven Test

Another alternative is to write a [table driven test](/testing/schema#table-driven-tests). This
test uses the `for_each` meta-argument, which accepts a map or a set of values and is used to generate
a test case for each item in the set or map.

Following similar logic to the example above, we will create _two_ table driven tests to
check each case:

1. A valid US Postal Code.
2. And invalid US Postal Code.

```hcl title="schema.hcl"

> test "schema" "us_postal_code_check_valid" {
>     parallel = true
>     for_each = [
>         {input = "12345", expected = "valid"},
>         {input = "12345-6789", expected = "valid"},
>     ]
>     log {
>         message = "Testing postal code: ${each.value.input} -> Expected: ${each.value.expected}"
>     }
>     exec {
>         sql = "SELECT '${each.value.input}'::us_postal_code"
>     }
> }

Reviewer: ```suggestion
test "schema" "us_postal_code_check_valid" {
  parallel = true
  for_each = [
    {input = "12345", expected = "valid"},
    {input = "12345-6789", expected = "valid"},
  ]
  log {
    message = "Testing postal code: ${each.value.input} -> Expected: ${each.value.expected}"
  }
  exec {
    sql = "SELECT '${each.value.input}'::us_postal_code"
  }
}
```

> test "schema" "us_postal_code_check_invalid" {
>     parallel = true
>     for_each = [
>         {input = "hello", expected = "invalid"},
>         {input = "123", expected = "invalid"},
>     ]
>     log {
>         message = "Testing postal code: ${each.value.input} -> Expected: ${each.value.expected}"
>     }
>     catch {
>         sql = "SELECT '${each.value.input}'::us_postal_code"
>     }
> }

Reviewer: ```suggestion
test "schema" "us_postal_code_check_invalid" {
  parallel = true
  for_each = [
    {input = "hello", expected = "invalid"},
    {input = "123", expected = "invalid"},
  ]
  log {
    message = "Testing postal code: ${each.value.input} -> Expected: ${each.value.expected}"
  }
  catch {
    sql = "SELECT '${each.value.input}'::us_postal_code"
  }
}
```

---
id: testing-domains

> title: Testing Database Domains

Reviewer: ```suggestion
title: Testing Postgres Domains
```

slug: /guides/testing/domains
---

Testing your database schema and migrations is crucial to ensure code behaves as expected, catch bugs early,
and prevent regressions. Databases enforce logic, constraints, and complex relationships,
so testing ensures these elements work correctly and remain intact after changes.

> In this guide we will learn how to use Atlas's [`schema test`](/testing/schema) command to test database domains.

Reviewer: ```suggestion
In this guide we will learn how to use Atlas's [`schema test`](/testing/schema) command to test [Postgres Domain Types](https://www.postgresql.org/docs/current/domains.html).
```

> ## Database Views

Reviewer: ?

> A `domain` is a user-defined data type that is based on an existing  data type but with optional constraints and d

Reviewer: ```suggestion
A `domain` is a user-defined data type that is based on an existing data type but with optional constraints and d
```

> efault values. Supported by PostgreSQL.

Reviewer: ```suggestion
efault values. 
```

:::info [Atlas Pro Feature](/features#pro-plan)

> Views are currently available only to [Atlas Pro users](/features#pro-plan). To use this feature, run:

Reviewer: Domain types

```
atlas login
```
:::

## Project Setup

### Schema File
For this example, let's assume we have the following schema, including a domain:

> 

Reviewer: ```suggestion
```
